ui, sql: surface Implicit Transactions for Statements

celiala opened the pull request:
Fixes: #36009.

This PR surfaces whether or not a statement was run in an implicit transaction for the web ui, so it's possible to infer (for example) whether or not the statement timing includes its replication latency (See #36009).

This involves:
- sql: Adding an `implicit_txn` flag to `node_statement_statistics` and `crdb_internal.node_statement_statistics`
- ui: surfacing `implicit_txn` on the Statement Details page
- ui: grouping statements by both statement and implicit_txn on both the Statement Details and Statements (Summary) pages

Statement Summary Page:
<img width="500" alt="Txns Summary" src="https://user-images.githubusercontent.com/3051672/59927193-e474ff00-9409-11e9-8415-ecb8e8a4197d.png">

Statement Summary, Tooltip:
<img width="400" alt="Txn Tooltip" src="https://user-images.githubusercontent.com/3051672/59953474-3b9cc300-944e-11e9-8311-e97e0c02165b.png">

Statement Details, Implicit Transaction:
<img width="300" alt="Txns Implicit" src="https://user-images.githubusercontent.com/3051672/59953507-6ab33480-944e-11e9-8978-8376196ea674.png">

Statement Details, Explicit Transaction:

<img width="300" alt="Txn Explicit" src="https://user-images.githubusercontent.com/3051672/59953506-6ab33480-944e-11e9-86ea-c98841dc2560.png">

Release note (admin ui change): Adding `Implicit Transaction` flag to Statements and Statement Details page.

jordanlewis committed:
sql: add implicit_txn flag to node_statement_statistics

Release note: None

celiala committed:
ui, sql: surface Implicit Transactions for Statements

This PR surfaces whether or not a statement was run in an implicit transaction
for the web ui, so it's possible to infer (for example) whether or not the
statement timing includes its replication latency (See #36009).

This involves:
- sql: Adding an `implicit_txn` flag to `node_statement_statistics` and
  `crdb_internal.node_statement_statistics`
- ui: surfacing `implicit_txn` on the Statement Details page
- ui: grouping statements by both statement and implicit_txn on both the
  Statement Details and Statements (Summary) pages

Example, Implicit Transaction is "Yes":

<img width="500" alt="Implicit Yes Summary" src="https://user-images.githubusercontent.com/3051672/59353920-0677ce80-8cf2-11e9-8b3a-5addb1cad11a.png">
<img width="1300" alt="Implicit Detail Yes" src="https://user-images.githubusercontent.com/3051672/59353919-0677ce80-8cf2-11e9-9442-7889a90a19b8.png">

Example, Implicit Transaction is "No":

<img width="500" alt="Implicit Summary No" src="https://user-images.githubusercontent.com/3051672/59353923-0677ce80-8cf2-11e9-80f4-0ceabdd9d98c.png">
<img width="1300" alt="Implicit Details No" src="https://user-images.githubusercontent.com/3051672/59353922-0677ce80-8cf2-11e9-82de-96af07e1a835.png">

Release note (admin ui change): Adding `Implicit Transaction` flag to
Statements and Statement Details page.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/36341)
<!-- Reviewable:end -->

vilterp commented on pkg/roachpb/app_stats.proto:
```diff
@@ -111,6 +111,7 @@ message StatementStatisticsKey {
   optional bool distSQL = 3 [(gogoproto.nullable) = false];
   optional bool failed = 4 [(gogoproto.nullable) = false];
   optional bool opt = 5 [(gogoproto.nullable) = false];
+  optional bool implicitTransaction = 6 [(gogoproto.nullable) = false];
```
nit: style in protos is `implicit_transaction`

nvanbenschoten commented:
@celiala thanks for writing this up. Any way I can help push it over the finish line?

celiala commented:
Thanks for the ping, Nathan: 
Jordan and I just sync'ed on how to finish this up -- should have an updated PR up soon.

celiala commented:
This is RFAL, thanks!

- Paired with Jordan to finish up PR #36340 & combine into this commit.
- Fixed proto naming nit.

celiala commented:
cc @Amruta-Ranade - we're planning to have this in the point release (19.1.1).

jordanlewis reviewed (commented):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

The tests are failing because I think the results of the query to `node_statement_statistics` can return in arbitrary order - to fix that, just add a `rowsort` declaration to the test. So it'd be `query TB colnames rowsort`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36341#-:-Lc1zC84-uKk6vTnOeeS:bmod8tu)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

celiala commented:
TFTR!

adding `rowsort` fixed the test, thx!

bors=r+

nvb reviewed (commented):
Thanks for fixing this!

The one thing I don't see is where we add a new column on the `/statements` page and where we ensure that the same statement executed sometimes in implicit transactions and sometimes in explicit transactions gets split out into two different statement types. Does this PR do that?

Also, I think the issue this claims to fix is wrong.

> Reviewed 13 of 13 files at r2, 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36341#-:-Lc7G1PU1HPVTN0yxSSN:b-64azz1)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @celiala)

___
*[pkg/sql/crdb_internal.go, line 595 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36341#-Lc7FaTY3TJXFwsuOgbQ:-Lc7FaTY3TJXFwsuOgbR:b-t37jn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee20a805f9b638f2a40f474f16421f7af3dbe0b/pkg/sql/crdb_internal.go#L595)):*
> ```Go
>   overhead_lat_avg    FLOAT NOT NULL,
>   overhead_lat_var    FLOAT NOT NULL,
>   implicit_txn        BOOL
> ```

This should be `NOT NULL`, right?


<!-- Sent from Reviewable.io -->

celiala commented:
TFTR!

> issue this claims to fix is wrong.

oops - fixed :)

> The one thing I don't see is where we add a new column on the /statements page 

ah, yes: we do need an `implicit txn` column on the statements page; will do.

> implicit_txn        BOOL

yes, should be `NOT NULL` - will fix.

nvanbenschoten commented:
I just noticed that this PR also fixes https://github.com/cockroachdb/cockroach/issues/36286.

celiala commented:
Hi all - this is RFAL, thanks!

* ui: Added a new column "Implicit Txn" on `/statements` page, to split these out into two different statement types
* fixed `implicit_txn` column on `crdb_internal.node_statement_statistics` to type `BOOL NOT NULL`

piyush-singh commented:
A small nit: I think this might be more readable if the column is instead "Transaction Type" ("Txn Type" works if we want to conserve space) and the rows are marked as "Explicit" or "Implicit". This way, when users scroll they won't need to scroll up to see whether the question for the column is for explicit or implicit. 

@Amruta-Ranade we should probably add a tooltip for the column header as well. My rough attempt to write up a tooltip: `Statements in explicit transactions are wrapped by BEGIN and COMMIT statements by the client. CockroachDB wraps individual statements in implicit transactions. Explicit transactions employ [transactional pipelining](https://www.cockroachlabs.com/docs/v19.1/architecture/transaction-layer.html#transaction-pipelining) and therefore report latencies that do not account for replication.`. Feel free to suggest any edits.

vilterp commented on pkg/ui/src/views/statements/statementsTable.tsx:
```diff
@@ -75,7 +77,13 @@ export function makeStatementsColumns(statements: AggregateStatistics[], selecte
+          implicitTxn={ stmt.implicitTxn }
+          app={ selectedApp }
+        />
+        ),
```
nit: outdent

vilterp commented on pkg/ui/src/views/statements/statementsTable.tsx:
```diff
@@ -129,6 +137,12 @@ function makeCommonColumns(statements: AggregateStatistics[])
   const latencyBar = latencyBarChart(statements);
 
   return [
+    {
```
I'd say while we're here, change the default sort column to latency (it's what people care about…)

vilterp commented on pkg/ui/src/views/statements/statementDetails.tsx:
```diff
@@ -344,6 +347,10 @@ class StatementDetails extends React.Component<StatementDetailsProps, StatementD
                 </td>
               </tr>
+              <tr className="numeric-stats-table__row--body">
+                <th className="numeric-stats-table__cell" style={{ textAlign: "left" }}>Transaction Type</th>
```
why does this column need to exist? won't it be the same for each row?

vilterp commented on pkg/ui/src/views/statements/statementsPage.tsx:
```diff
@@ -59,7 +69,7 @@ class StatementsPage extends React.Component<StatementsPageProps & RouteProps, S
     this.state = {
       sortSetting: {
-        sortKey: 1,
+        sortKey: 2,
```
Honestly, I think latency should be the default sort (it's what people care about; I don't think "total time" is that meaningful…)

celiala commented:
TFTR!

Updates:
- Statement Summary/Details page: Latency is default sorted column
- Statement Details: removed Txn Type column
- Statement Summary: Txn Type column: addded Tooltip (see screenshot)
- fixed outdent nit

Amruta-Ranade commented:
@piyush-singh @celiala  I need help understanding the context for the tooltip - can we sync on Monday to figure it out?

celiala commented:
bors r+

@Amruta-Ranade and I synced offline: I'll merge feature now, so that this gets into the alpha cut. And we'll sync on context sometime this week, to adjust copy/documentation as needed for next cut.
